akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -447,33 +444,28 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
     startTransport()
     topLevelFlightRecorder.loFreq(Transport_Started, NoMetaData)
 
-    val udp = settings.Transport == ArterySettings.AeronUpd
-    val port =
-      if (settings.Canonical.Port == 0) {
-        if (settings.Bind.Port != 0) settings.Bind.Port // if bind port is set, use bind port instead of random
-        else ArteryTransport.autoSelectPort(settings.Canonical.Hostname, udp)
-      } else settings.Canonical.Port
-
-    _localAddress = UniqueAddress(
-      Address(ArteryTransport.ProtocolName, system.name, settings.Canonical.Hostname, port),
-      AddressUidExtension(system).longAddressUid)
-    _addresses = Set(_localAddress.address)
-
-    // TODO: This probably needs to be a global value instead of an event as events might rotate out of the log
-    topLevelFlightRecorder.loFreq(Transport_UniqueAddressSet, _localAddress.toString())
-
     materializer = ActorMaterializer.systemMaterializer(settings.Advanced.MaterializerSettings, "remote", system)
     controlMaterializer =
       ActorMaterializer.systemMaterializer(settings.Advanced.ControlStreamMaterializerSettings, "remoteControl", system)
 
     messageDispatcher = new MessageDispatcher(system, provider)
     topLevelFlightRecorder.loFreq(Transport_MaterializerStarted, NoMetaData)
 
-    val boundPort = runInboundStreams()
+    val (port, boundPort) = bindInboundStreams()
+
+    _localAddress = UniqueAddress(
+      Address(ArteryTransport.ProtocolName, system.name, settings.Canonical.Hostname, port),
+      AddressUidExtension(system).longAddressUid)
+    _addresses = Set(_localAddress.address)
+
     _bindAddress = UniqueAddress(
       Address(ArteryTransport.ProtocolName, system.name, settings.Bind.Hostname, boundPort),
       AddressUidExtension(system).longAddressUid)
 
+    topLevelFlightRecorder.loFreq(Transport_UniqueAddressSet, _localAddress.toString())
+
+    runInboundStreams(port, boundPort)
```

Comment:
I guess `runInboundStreams` can only be run after `_localAddress` is known? Is that the main reason for the split?

Reply:
Yes, exactly: when building the `Flow` that is materialized for each incoming connection already the `_localAddress` is passed in in several places.